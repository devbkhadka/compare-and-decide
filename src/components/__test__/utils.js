import React from 'react'
import { Provider } from 'react-redux'
import { IntlProvider } from 'react-intl'
import { render } from '@testing-library/react'
import configureMockStore from 'redux-mock-store'

const middlewares = []
const mockStore = configureMockStore(middlewares)

export function renderWithIntl(component) {
    return render(<IntlProvider locale="en" messages={{}}>{ component }</IntlProvider>)
}

export function renderWithProvider(component, initialState={}) {
    const store = mockStore(initialState)
    return [render(<Provider store={store}>
        <IntlProvider locale="en" messages={{}}>{ component }</IntlProvider>
    </Provider>), store]
}