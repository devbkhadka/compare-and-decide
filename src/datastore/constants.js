export const ADD_ITEM_WITH_TITLE = 'ADD_ITEM_WITH_TITLE'