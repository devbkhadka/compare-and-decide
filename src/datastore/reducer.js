import { ADD_ITEM_WITH_TITLE } from './constants'

const initialState = {
    items:[]
}

export default function (state=initialState, payload) {
    switch(payload.type) {
        case ADD_ITEM_WITH_TITLE:
            if(payload.data){
                return {...state, items: [...state.items, payload.data]}
            }
        
        default:
            return state
    }
}