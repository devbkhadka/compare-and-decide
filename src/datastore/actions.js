import { UPDATE_ITEMS, 
    UPDATE_LANGUAGE, 
    UPDATE_ATTRIBUTES,
    UPDATE_PREFERRED_ITEM
} from './constants'

import {COMPARISION_STATUS} from './reducer'

export const ITEMS_KEY = 'items-key'
export const ATTRIBUTES_KEY = 'attributes-key'
export const PREFETED_ITEM_TITLE_KEY = 'prefered-item-key'

export function addItemWithTitle(title) {
    return (dispatch, getState) => {

        const state = getState()
        if(!title || state.items[title]) {
            return
        }

        const item = {
            title: title,
            values: {}
        }
        state.items[item.title] = item
        updateAndSaveItems(state.items, dispatch)
    }
}

export function loadStateFromStorage() {
    return dispatch => {
        const itemsJson = localStorage.getItem(ITEMS_KEY)
        if(itemsJson) {
            dispatch(updateItems(JSON.parse(itemsJson)))
        }

        const attributesJson = localStorage.getItem(ATTRIBUTES_KEY)
        if(attributesJson) {
            dispatch(updateAttributes(JSON.parse(attributesJson)))
        }

        const preferedItemTitle = localStorage.getItem(PREFETED_ITEM_TITLE_KEY)
        if(preferedItemTitle){
            dispatch(updatePreferedItemTitle(preferedItemTitle))
        }
    }
}

export function updateItems(items) {
    return {
        type: UPDATE_ITEMS,
        data: items
    }
}

export function updateLanguage(language) {
    return {
        type: UPDATE_LANGUAGE,
        data: language
    }
}

export function addAttribute(attribute) {
    return (dispatch, getState)=>{
        const attributes = getState().attributes
        if(attribute && attributes.indexOf(attribute)===-1) {
            attributes.push(attribute)

            updateAndSaveAttributes(attributes, dispatch)
        }
    }
}

export function updateAttributes(attributes) {
    return {
        type: UPDATE_ATTRIBUTES,
        data: attributes
    }
}

export function updateItemAttributeValue(item, attr, value) {
    return (dispatch, getState) => {
        const state = getState()
        const newItems = {...state.items}
        newItems[item.title] = {...item, values: {...item.values, [attr]: value}}
        
        updateAndSaveItems(newItems, dispatch)
    }
}

export function setPreferedItemTitle(itemTitle) {
    return (dispatch, getState) => {
        const state = getState()
        const curPreferedTitle = state.preferedItemTitle

        const item = state.items[itemTitle]
        if(!item) {
            // TODO: This branch not tested
            return
        }
        if(curPreferedTitle && state.items[curPreferedTitle]){
            // TODO: This branch not tested
            state.items[curPreferedTitle].comparisionStatus = undefined
        }
        item.comparisionStatus = COMPARISION_STATUS.ITEM_PREFERED

        updateAndSaveItems(state.items, dispatch)
        updateAndSavePreferedItemTitle(itemTitle, dispatch)
        
    }
}

export function updatePreferedItemTitle(itemTitle) {
    return {
        type: UPDATE_PREFERRED_ITEM,
        data: itemTitle
    }
}

export function rejectItem(itemTitle){
    return (dispatch, getState) => {
        const state = getState()
        state.items[itemTitle].comparisionStatus = COMPARISION_STATUS.ITEM_REJECTED

        updateAndSaveItems(state.items, dispatch)
    }
}

export function deleteAttribute(attribute) {
    return (dispatch, getState) => {
        const {items, attributes} = getState()
        const newAttributes = attributes.filter(attr=>attribute !== attr)

        for(const key in items) {
            const values = items[key].values
            delete values[attribute]
        }

        updateAndSaveItems(items, dispatch)
        updateAndSaveAttributes(newAttributes, dispatch)
    }
}

export function deleteItem(itemTitle) {
    return (dispatch, getState) => {
        const {items} = getState()
        delete items[itemTitle]

        updateAndSaveItems(items, dispatch)
    }
}

export function clearComparisionStatus(itemTitle) {
    return (dispatch, getState) => {
        const {items, preferedItemTitle} = getState()
        items[itemTitle].comparisionStatus = undefined
        updateAndSaveItems(items, dispatch)
        if(preferedItemTitle===itemTitle) {
            updateAndSavePreferedItemTitle(undefined, dispatch)
        }
    }
}


function updateAndSaveItems(items, dispatch) {
    localStorage.setItem(ITEMS_KEY, JSON.stringify(items))
    dispatch(updateItems(items))
}

function updateAndSavePreferedItemTitle(itemTitle, dispatch) {
    localStorage.setItem(PREFETED_ITEM_TITLE_KEY, itemTitle)
    dispatch(updatePreferedItemTitle(itemTitle))
}

function updateAndSaveAttributes(attributes, dispatch) {
    localStorage.setItem(ATTRIBUTES_KEY, JSON.stringify(attributes))
    dispatch(updateAttributes(attributes))
}