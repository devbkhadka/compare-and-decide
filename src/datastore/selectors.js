import { COMPARISION_STATUS } from './reducer'

const itemsSelector = (items, selector)=>{
    const selectedItems = []
    for(const key of Object.keys(items)){
        const item = items[key]
        if(selector(item)){
            selectedItems.push(item)
        }
    }
    return selectedItems
}

export const remainingItemsSelector = state=> itemsSelector(state.items, item=>!item.comparisionStatus)
export const rejectedItemsSelector = state =>
    itemsSelector(state.items, item=> item.comparisionStatus===COMPARISION_STATUS.ITEM_REJECTED)