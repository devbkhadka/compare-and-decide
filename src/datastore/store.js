import {createStore, applyMiddleware, compose} from 'redux'
import ReduxThunk from 'redux-thunk'
import reducer from './reducer'


export const middleWares = [ReduxThunk]

const devToolEnhancer = (window && window.__REDUX_DEVTOOLS_EXTENSION__ && 
                            window.__REDUX_DEVTOOLS_EXTENSION__())
let enhancer = applyMiddleware(...middleWares)
if(devToolEnhancer) {
    enhancer = compose(enhancer, devToolEnhancer)
} 

const store = createStore(reducer, enhancer)

export default store
