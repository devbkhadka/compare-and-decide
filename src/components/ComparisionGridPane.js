import React from 'react'
import {Box, makeStyles, Grid, Hidden, Typography, IconButton} from '@material-ui/core'
import CheckCircleIcon from '@material-ui/icons/CheckCircle';
import Title from './shared/Title'
import { useSelector, useDispatch } from 'react-redux'
import { remainingItemsSelector } from '../datastore/selectors'
import { setPreferedItemTitle, rejectItem } from '../datastore/actions';

const useStyles = makeStyles((theme)=>({
    root: {
        position: 'relative',
        height: '100%'
    },
    separator: {
        height: '100%',
        '& > .seperatorBorder': {
            borderRight: `1px solid ${theme.palette.divider}`,
        } 
    },
    content: {
        position: 'absolute'
    },
    attribute: {
        fontWeight: 'bold'
    },
    columnHeader: {
        display: 'flex'
    }
    
}))

const ComparisionGridPane = ()=> {
    const classes = useStyles()
    const attributes = useSelector(state=>state.attributes)
    const preferedItem = useSelector(state=>state.items[state.preferedItemTitle])
    const otherItem = useSelector(state=>remainingItemsSelector(state)[0])
    const dispatch = useDispatch()

    const selectPreferedItem = (item)=> {
        const rejected = item===preferedItem? otherItem: preferedItem
        dispatch(setPreferedItemTitle(item.title))
        dispatch(rejectItem(rejected.title))
    }

    return (!!preferedItem) ? <Box data-testid="root-content" className={classes.root}>
        <Grid container spacing={2} className={classes.content}>
            {!otherItem ? 
                <Grid item xs={12}>
                    <Title>You have choosen "{preferedItem.title}"</Title>
                </Grid>
                :
                <>
                    <Hidden xsDown>
                        <Grid item xs={12} sm={!!otherItem?2:4}>
                            <Title>Attributes</Title>
                        </Grid>
                    </Hidden>
                    <Grid item sm={!!otherItem?5:8} xs={12} className={classes.columnHeader} >
                        <Title>{preferedItem.title}</Title> 
                        <IconButton data-testid='btnCheck' onClick={()=>selectPreferedItem(preferedItem)}>
                            <CheckCircleIcon fontSize='small'/>
                        </IconButton>
                    </Grid>
                </>
            }
            {!!otherItem && 
                <Grid item sm={5} xs={12} className={classes.columnHeader}>
                    <Title >{otherItem.title}</Title>
                    <IconButton data-testid='btnCheck' onClick={()=>selectPreferedItem(otherItem)}>
                        <CheckCircleIcon fontSize='small'/>
                    </IconButton>
                </Grid>
            }
            {attributes.map((attr,i)=>(
                <React.Fragment key={`${attr}-${i}`}>
                    <Grid item xs={12} sm={!!otherItem?2:4} className={classes.attribute}>
                        <Typography variant="body1">{attr}</Typography>
                    </Grid>
                    <Grid item xs={6} sm={!!otherItem?5:8}>
                        <Typography variant="body1">
                            {preferedItem.values && preferedItem.values[attr]}
                        </Typography>
                    </Grid>
                    {!!otherItem && 
                        <Grid item xs={6} sm={5}>
                            <Typography variant="subtitle1">
                                {otherItem.values && otherItem.values[attr]}
                            </Typography>
                        </Grid>
                    }
                </React.Fragment>
            ))}

        </Grid>
        <Hidden xsDown>
            {!!otherItem && 
                <Grid container className={classes.separator}>
                    <Grid item sm={2} />
                    <Grid item sm={5} className='seperatorBorder'/>
                    <Grid item sm={5} />
                </Grid>
            }
        </Hidden>
    </Box> : 
    <Typography variant='body1'>Please select a prefered item from left pane to start comparision</Typography>
}

export default ComparisionGridPane