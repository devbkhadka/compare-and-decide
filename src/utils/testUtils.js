import React from 'react'
import { Provider } from 'react-redux'
import { IntlProvider } from 'react-intl'
import { render } from '@testing-library/react'
import configureMockStore from 'redux-mock-store'
import {middleWares} from '../datastore/store'
import ReduxThunk from 'redux-thunk'


const mockStore = configureMockStore(middleWares)

export function renderWithIntl(component) {
    return render(<IntlProvider locale="en" messages={{}}>{ component }</IntlProvider>)
}

export function renderWithProvider(component, initialState={language: 'en'}) {
    const store = mockStore(initialState)
    return [render(<Provider store={store}>
        <IntlProvider locale="en" messages={{}}>{ component }</IntlProvider>
    </Provider>), store]
}