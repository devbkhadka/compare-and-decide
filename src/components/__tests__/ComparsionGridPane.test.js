import React from 'react'

import ComparisionGridPane from '../ComparisionGridPane'
import { renderWithProvider } from '../../utils/testUtils'
import {remainingItemsSelector as mockRemainingItemsSelector} from '../../datastore/selectors'
import { within, fireEvent } from '@testing-library/react'
import {
    setPreferedItemTitle as mockSetPreferedItemTitle,
    rejectItem as mockRejectItem
} from '../../datastore/actions'
import { COMPARISION_STATUS } from '../../datastore/reducer'

jest.mock('../../datastore/actions')
mockSetPreferedItemTitle.mockImplementation(()=>()=>null)
mockRejectItem.mockImplementation(()=>()=>null)
jest.mock('../../datastore/selectors')

jest.spyOn(global, 'confirm')

describe('ComparisionGridPane Component', ()=>{
    const copyBaseState = ()=>{
        const state = {
            items:{
                Item1: {
                    title: 'Item1',
                    values: {
                        attr1: 'Item1-Value1',
                        attr2: 'Item1-Value2',
                        attr3: 'Item1-Value3'
                    }
                },
                Item2: {
                    title: 'Item2',
                    values: {
                        attr2: 'Item2-Value2',
                        attr3: 'Item2-Value3'
                    }
                },
                Item3: {
                    title: 'Item3',
                    values: {
                        attr1: 'Item3-Value1',
                        attr2: 'Item3-Value2',
                        attr3: 'Item3-Value3'
                    }
                },
                Item4: {
                    title: 'Item4',
                    values: {
                        attr3: 'Item4-Value3'
                    }
                }
            },
            attributes: ['attr1', 'attr2', 'attr3']
        }
        return JSON.parse(JSON.stringify(state))
    }
    describe('When prefered item not selected', ()=>{
        const state = copyBaseState()
        const {Item2, Item3, Item4} = state.items
        beforeEach(()=>{
            mockRemainingItemsSelector.mockImplementation(()=>{
                return [Item2, Item3, Item4]
            }).mockName('nonEmptyRemainingItemSelector')
        })
        it('Should show message to select prefered item', ()=>{
            const [rendered] = renderWithProvider(<ComparisionGridPane/>, copyBaseState())
            rendered.getByText(/Please select a prefered item.*/)
            expect(rendered.queryByTestId('root-content')).toBeNull()
        })
    })
        

    describe('When prefered item is selected and items are remaining', ()=>{
        const state = copyBaseState()
        const {Item2, Item3, Item4} = state.items
        state.preferedItemTitle = Item2.title
        beforeEach(()=>{
            mockRemainingItemsSelector.mockImplementation(()=>{
                return [Item3, Item4]
            }).mockName('nonEmptyRemainingItemSelector')
        })
        
        it('Should show comparision between prefered Item and 1st Item from remaining', ()=>{
            const [rendered] = renderWithProvider(<ComparisionGridPane/>, state)
            expect(rendered.queryAllByText(/Item2-Value\d/).length).toBe(2)
            expect(rendered.queryAllByText(/Item3-Value\d/).length).toBe(3)
        })

        describe("Select prefered item", ()=>{
            let btnForPrefered
            let btnForOther
            beforeEach(()=>{
                const [rendered] = renderWithProvider(<ComparisionGridPane/>, state)
    
                const preferedTitle = rendered.getByText(Item2.title)
                const otherTitle = rendered.getByText(Item3.title)
    
                btnForPrefered = within(preferedTitle.parentElement).getByTestId('btnCheck')
                btnForOther = within(otherTitle.parentElement).getByTestId('btnCheck')
            })

            afterEach(()=>{
                mockSetPreferedItemTitle.mockClear()
                mockRejectItem.mockClear()
            })
    
            it("Should be able to choose prefered item again as prefered", ()=>{
                global.confirm.mockImplementation(()=>true)
                fireEvent.click(btnForPrefered)
                expect(mockSetPreferedItemTitle).toBeCalledWith(Item2.title)
                expect(mockRejectItem).toBeCalledWith(Item3.title)
            })

            it("Should be able to choose other item as prefered", ()=>{
                global.confirm.mockImplementation(()=>true)
                fireEvent.click(btnForOther)
                expect(mockSetPreferedItemTitle).toBeCalledWith(Item3.title)
                expect(mockRejectItem).toBeCalledWith(Item2.title)
            })

            it("Should be able to cancel selection after clicking item", ()=>{
                global.confirm.mockImplementation(()=>false)
                fireEvent.click(btnForPrefered)
                expect(mockSetPreferedItemTitle).not.toBeCalled()
            })
        })
        
        // Should be able to add comment on the item
        
    })

    
    describe("When prefered item is selected and all comparision finished", ()=>{
        const state = copyBaseState()
        const {Item2, Item3, Item4} = state.items
        state.preferedItemTitle = Item2.title
        Item3.comparisionStatus = COMPARISION_STATUS.ITEM_REJECTED
        Item4.comparisionStatus = COMPARISION_STATUS.ITEM_REJECTED

        beforeEach(()=>{
            mockRemainingItemsSelector.mockImplementation(()=>{
                return []
            }).mockName('emptyRemainingItemsSelector')
        })
        
        it("should show last prefered item as selection", ()=>{
            const [rendered] = renderWithProvider(<ComparisionGridPane/>, state)
            expect(rendered.baseElement).toMatchSnapshot()
        })
    })
        
})