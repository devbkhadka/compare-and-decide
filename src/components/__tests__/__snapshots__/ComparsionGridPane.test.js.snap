// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComparisionGridPane Component When prefered item is selected and all comparision finished should show last prefered item as selection 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root MuiBox-root-558 makeStyles-root-552"
      data-testid="root-content"
    >
      <div
        class="MuiGrid-root makeStyles-content-554 MuiGrid-container MuiGrid-spacing-xs-2"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <h5
            class="MuiTypography-root Component-root-662 MuiTypography-h5"
          >
            You have choosen "Item2"
          </h5>
        </div>
        <div
          class="MuiGrid-root makeStyles-attribute-555 MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4"
        >
          <h6
            class="MuiTypography-root MuiTypography-subtitle1"
          >
            attr1
          </h6>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-md-8"
        >
          <p
            class="MuiTypography-root MuiTypography-body1"
          />
        </div>
        <div
          class="MuiGrid-root makeStyles-attribute-555 MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4"
        >
          <h6
            class="MuiTypography-root MuiTypography-subtitle1"
          >
            attr2
          </h6>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-md-8"
        >
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            Item2-Value2
          </p>
        </div>
        <div
          class="MuiGrid-root makeStyles-attribute-555 MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4"
        >
          <h6
            class="MuiTypography-root MuiTypography-subtitle1"
          >
            attr3
          </h6>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-md-8"
        >
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            Item2-Value3
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;
